### ISSUE TITLE
 Jinja2 filter operator precedence is confusing

### ISSUE THREAD

**Issue by pfalcon** [pallets/jinja#379]
I posted this to mailing list (https://groups.google.com/forum/#!topic/pocoo-libs/xR5dHsLiHj8) initially, but didn't get much response (yet), so maybe here it will make more sense.

I'm trying to figure how expression like "1 + 2 | float" is supposed to be evaluated: as "(1 + 2) | float" or "1 + (2 | float)". Let me start with saying that it's intuitively clear that Jinja2's pipe tries to base on unix shell pipe, but then it's just intuitive that | operator should have low priority - first an entire expression ("a subcommand with all args" in unix shell land) is evaluated, then it's piped thru filter(s).

But reading thru docs, http://jinja.pocoo.org/docs/dev/templates/#filters doesn't clarify that explicitly. It talks about "variables can be modified by filters", but what would that mean? Either "variable" is just a colloquial word, which rather should be "expression", but then its precedence is unclear. However, if "variable" is a technical term, then it implies (just implies, doesn't make explicit) that | binds tightly to primary expressions (because still, foo.bar isn't a "variable", but more a "primary expression"). But then does that section imply that filters cannot be applied to constant values? (because constants are clearly not variables)

All in all, that intro section on filters (http://jinja.pocoo.org/docs/dev/templates/#filters) leaves impression of some magicity of pipe operator. However, later, in http://jinja.pocoo.org/docs/dev/templates/#other-operators shade of magicity is gone - it's nothing but an operator among bunch of others. Except that whole "Expressions" section doesn't say anything about relative precedence of different operators.

I'm not the only one who gets confused by this matter - see for example http://stackoverflow.com/a/23249743/496009 which starts with "The Jinja documentation doesn't specify any precedence rules".

So, may I suggest following documentation changes: 1) "Expressions" section to be augment with operator precedence table/list; 2) "Filters" intro section to mention precedence explicitly either, perhaps by giving an example that "a + b | filter" binds as "a + (b | filter)" and not "(a + b) | filter" ? Thanks in advance.

Of related question, is there a jinja2 native way to test rendering of a template - for example, a command-line utility which would be installed by pip/OS packages and allowed to quickly test a template with particular (simple) variable bindings, for example "jinja2-render template.tpl foo=1 bar='string'"?

**Comment by candlerb** [comment]
I am guessing that jinja2 took its operator precedence rules directly from python. This would be reasonable, except that the pipe operator is used to apply a filter, not to do a bitwise-OR of its two operands.

**Comment by Digenis** [comment]
If it takes the precedence from python,
this clarification can fit in a sentence.
In my opinion, keeping the documentation shorter by a sentence
doesn't justify letting in such an ambiguity.

**Comment by untitaker** [comment]
I fully agree with documenting the current behavior better, however, any changes to this behavior won't be possible anytime soon due to backwards compat concerns.

The latest talk at PiterPy by @mitsuhiko clarified that things like this were a mistake.

**Comment by ngaya-ll** [comment]
@candlerb This is not a correct supposition.  Jinja2 has its own built-in parser, so precedence is independent of Python.

@pfalcon Filters and tests have higher precedence than binary operators.  For example `{{ [3] + [2,1] | sort }}` yield `[3, 1, 2]`, not `[1, 2, 3]`.

**Comment by jeffwidman** [comment]
See also #119 which mentions a very similar issue.

**Comment by gggeek** [comment]
Sorry to come off as a troll, but 3 years have passed, and still the official doc does not list the operators precedence...
Surely adding a couple of paragraphs to documentation would take less time than fixing the parser so that it follows the same precedence rules as python does?

**Comment by jmehnle** [comment]
@gggeek, this is a good idea. Are you able to create a pull request that adds this to the docs?

**Comment by gggeek** [comment]
@jmehnle sorry, I am totally swamped with work at the moment :-(

**Comment by artificial-intelligence** [comment]
Hi!

So, when searching for precedence handling in jinja via a search engine I ended up here, which thankfully has the information I'm looking for (thanks @ngaya-ll ), but at least I can find this kind of information reflected in the actual jinja docs, where I would expect something like this.

Being an opensource maintainer myself I understand stuff takes time and sometimes falls through the cracks.

But would someone mind maybe documenting this, maybe with a short sentence?

If you are happy to accept drive-by contributions and have some guidance maybe I could also provide a PR myself, but I'm not familiar with the project structure at all until now. I'm really just a happy user, so that could also take some time and hand holding and maybe a little sentence is easier to add than to interact with a drive-by contributor?

Anyway, thanks if someone is working on this (maybe I also just overlooked where this is documented, but I can't really find it, so also thanks if someone can point me in the right direction)!
